### PR TITLE
gh-113666: Adding missing UF_ and SF_ flags to module 'stat'

### PR DESCRIPTION
**Pull request by ronaldoussoren**
Add some constants to module 'stat' that are used on macOS.

* Issue: gh-113666

----
📚 Documentation preview 📚: https://cpython-previews--113667.org.readthedocs.build/

### CODE REVIEW

**Comment by ronaldoussoren** [comment]
There will be an update to this PR, just noticed that there's also a C extension with a definitions.

File: Lib/stat.py [hunk 1]
```diff
@@ -111,18 +112,31 @@ def S_ISWHT(mode):
...
 UF_IMMUTABLE = 0x00000002  # file may not be changed
 UF_APPEND    = 0x00000004  # file may only be appended to
 UF_OPAQUE    = 0x00000008  # directory is opaque when viewed through a union stack
 UF_NOUNLINK  = 0x00000010  # file may not be renamed or deleted
-UF_COMPRESSED = 0x00000020 # OS X: file is hfs-compressed
-UF_HIDDEN    = 0x00008000  # OS X: file should not be displayed
+UF_COMPRESSED = 0x00000020 # macOS: file is compressed
+UF_TRACKED   = 0x00000040  # macOS: used for handling document IDs
+UF_DATAVAULT = 0x00008000  # macOS: entitlement needed for I/O
+UF_HIDDEN    = 0x00008000  # macOS: file should not be displayed
```
Comment (serhiy-storchaka) [inline Lib/stat.py]: Why are they equal?

File: Lib/stat.py [hunk 2]
```diff
@@ -111,18 +112,31 @@ def S_ISWHT(mode):
 
 # Names for file flags
 
+if sys.platform == "darwin":
+    # Group of bits in st_flags, specific to Apple platforms.
+
+    UF_SETTABLE = 0xffff      # owner settable flags
+    SF_SUPPORTED = 0x9f0000   # superuser supported flags
```
Comment (serhiy-storchaka) [inline Lib/stat.py]:
```suggestion
    UF_SETTABLE  = 0x0000ffff # owner settable flags
    SF_SUPPORTED = 0x009f0000 # superuser supported flags
```

File: Lib/stat.py [hunk 3]
```diff
@@ -111,18 +112,31 @@ def S_ISWHT(mode):
...
+    SF_SETTABLE  = 0x3ffff000 # superuser settable flags
```
Comment (serhiy-storchaka) [inline Lib/stat.py]: It is suspicious that `UF_SETTABLE & SF_SETTABLE != 0`.

File: Lib/stat.py (hunk 1 above)
Comment (ronaldoussoren) [reply Lib/stat.py]: Probably because I'm bad at copy&paste :-(

**Comment by ronaldoussoren** [comment]
Something I noticed while looking at stat.py & _stat.c is that the two contain a lot of the same definitions. 

IMHO it would be better to strip duplicates (DRY and all), and possibly just drop stat.py (e.g. rename the `_stat` extension module to `stat`). 

I also don't particularly like defining names that aren't relevant for the current platform (for example `S_IFDOOR` which AFAIK is only relevant for Solaris). Changing that is something for a different PR though, this would change the interface for `stat` and requires the usual deprecation cycle before removing.

File: Lib/stat.py (hunk 3 above)
Comment (ronaldoussoren) [reply Lib/stat.py]: I intend to add a test for macOS that tests against the actual platform values for every constant in `stat`, even if that might not survive review.

Comment (ronaldoussoren) [reply Lib/stat.py]: > It is suspicious that `UF_SETTABLE & SF_SETTABLE != 0`.

This was a copy&paste error as well.

**Comment by ronaldoussoren** [comment]
I've updated the PR:
- Fixed typos noted by @serhiy-storchaka 
- Added the new constants to _stat.c as well
- Added a test case that tests the values on macOS (this is a bit overkill, but makes it easier to validate the changes)

I still don't like having the definitions in two places and exporting definitions not relevant for the current platform (e.g. `S_IFDOOR` on anything that isn't Solaris), I'll file a separate issue about those.

File: Doc/library/stat.rst [hunk 4]
```diff
@@ -374,10 +382,46 @@ The following flags can be used in the *flags* argument of :func:`os.chflags`:
...
 
+.. data:: UF_SUPPORTED
+
+   All super-user supported flags
+
+   .. availability:: macOS
+
+   .. versionadded: 3.13
+
+.. data:: UF_SETTABLE
```
Comment (serhiy-storchaka) [inline Doc/library/stat.rst]:
```suggestion
.. data:: SF_SETTABLE
```

**Comment by serhiy-storchaka** [comment]
I thought about adding some general tests, like testing that UF_SETTABLE, SF_SETTABLE and SF_SYNTHETIC do not intersect, SF_SUPPORTED is a subset of SF_SETTABLE, all `UF_*` flags are in UF_SETTABLE, and all `SF_*` flags are in SF_SETTABLE or SF_SYNTHETIC, etc. But your tests are more strict.

I searched for SF_SETTABLE and found it in the xnu sources (https://opensource.apple.com/source/xnu/xnu-2050.18.24/bsd/sys/stat.h), where it's definition is different. Also, there are no SF_SUPPORTED and SF_SYNTHETIC there. Does it matter?

Some flags are defined in `stat.py` and `_stat.c` because we want to provide some values even if the platform do not support them. But values from `_stat.c` override default values in `stat.py`. If meta-flags are only provided on macOS, I think that there is no need to define default values for them in `stat.py`.

**Comment by ronaldoussoren** [comment]
> I thought about adding some general tests, like testing that UF_SETTABLE, SF_SETTABLE and SF_SYNTHETIC do not intersect, SF_SUPPORTED is a subset of SF_SETTABLE, all `UF_*` flags are in UF_SETTABLE, and all `SF_*` flags are in SF_SETTABLE or SF_SYNTHETIC, etc. But your tests are more strict.

My tests are basically written by copying the `#define`-s from Apple headers (macOS 14.2 SDK) and transforming those to test assertions.

> 
> I searched for SF_SETTABLE and found it in the xnu sources (https://opensource.apple.com/source/xnu/xnu-2050.18.24/bsd/sys/stat.h), where it's definition is different. Also, there are no SF_SUPPORTED and SF_SYNTHETIC there. Does it matter?

Those definitions are old, I'm not entirely sure when SF_SYNTHETIC was introduced, but likely when the APFS filesystem was added (10.13), or when they added the system volume and firm links (10.15)

See https://opensource.apple.com/source/xnu/xnu-7195.81.3/bsd/sys/stat.h.auto.html for a more recent copy.

The difference between the older and newer versions shouldn't be a problem, all bits that are set in the older SF_SETTABLE but not in the newer have no assigned meaning in the older release. 

> 
> Some flags are defined in `stat.py` and `_stat.c` because we want to provide some values even if the platform do not support them. But values from `_stat.c` override default values in `stat.py`. If meta-flags are only provided on macOS, I think that there is no need to define default values for them in `stat.py`.

I'll do some more research, google seems to indicate that SF_SETTABLE and SU_SETTABLE are available in at least glibc as well (with the older definition you found), in which case:

- Add SU_SETTABLE and SF_SETTABLE to stat.py with the Linux definition, and remove the "if darwin" block
- Add SF_SUPPORTED and SF_SYNTHETIC only in stat.c (and override SF_SETTABLE there for macOS by way of using the system definition).

**Comment by ronaldoussoren** [comment]
I didn't find UF_SETTABLE and SF_SETTABLE on Linux, but did find them on FreeBSD. The PR therefore handles them like other generic flags (always available).

A slight complication: As you noticed older versions of macOS have a definition of SF_SETTABLE that doesn't match the definition in the current SDK, and "older" includes at least macOS 10.9 which is the oldest version  of the OS targeted by our installers. 

To end up with an interface that's easily explainable I've chooses to redefine SF_SETTABLE on those older macOS versions, and define SF_SUPPORTED and SF_SYNTHETIC there as well. That way we can describe those as being available on macOS, without having to explain that availability and values depend on how the installation was build (old SDK vs. new SDK).   This should affect usage of these attributes because the extra bits set in the older definition of SF_SETTABLE do not have a defined meaning on those older releases of macOS.

File: Lib/test/test_stat.py [hunk 5]
```diff
@@ -241,6 +241,18 @@ def test_module_attributes(self):
             self.assertTrue(callable(func))
             self.assertEqual(func(0), 0)
 
+    def test_flags_consistent(self):
+        self.assertFalse(self.statmod.UF_SETTABLE & self.statmod.SF_SETTABLE)
+
+        for flag in self.file_flags:
+            if flag.startswith("UF"):
+                self.assertTrue(getattr(self.statmod, flag) & self.statmod.UF_SETTABLE, f"{flag} notin UF_SETTABLE")
```
Comment (serhiy-storchaka) [inline Lib/test/test_stat.py]:
```suggestion
                self.assertTrue(getattr(self.statmod, flag) & self.statmod.UF_SETTABLE, f"{flag} not in UF_SETTABLE")
```

File: Modules/_stat.c [hunk 6]
```diff
@@ -579,6 +604,7 @@ stat_exec(PyObject *module)
     ADD_INT_MACRO(module, UF_TRACKED);
     ADD_INT_MACRO(module, UF_DATAVAULT);
     ADD_INT_MACRO(module, UF_HIDDEN);
+    ADD_INT_MACRO(module, SF_SETTABLE);
```
Comment (serhiy-storchaka) [inline Modules/_stat.c:607]: #ifdef SF_SETTABLE

File: Modules/_stat.c [hunk 7]
```diff
@@ -226,10 +230,22 @@ typedef unsigned short mode_t;
...
+#  define UF_DATAVAULT 0x00000080
+#endif
+
 #ifndef UF_HIDDEN
 #  define UF_HIDDEN 0x00008000
 #endif
 
+#ifndef SF_SETTABLE
+#  define SF_ARCHIVED 0xffff0000
+#endif
```
Comment (serhiy-storchaka) [inline Modules/_stat.c]: Why is this needed?

Comment (ronaldoussoren) [reply Modules/_stat.c]: Sigh, I've check the commit a couple of times and still missed that I forgot to update the name here.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

I only wondering why define default values for these flags both in `_stat.c` and `stat.py`? Usually, the C implementation only expose constants if they are defined:
```c
#ifdef SOMETHING
    ADD_INT_MACRO(module, SOMETHING);
#endif
```
Then the Python implementation can define the default value and override it with import from the C module if it is defined, or keep it undefined.
```py
SOMETHING = ...
try:
    from _somemodule import SOMETHING
except ImportError:
    pass
```
or
```py
try:
    from _somemodule import SOMETHING
except ImportError:
    SOMETHING = ...
```
Defining it in the C code if it is undefined is pretty unusual.

**Comment by ronaldoussoren** [comment]
> LGTM.
> 
[...]
> 
> Defining it in the C code if it is undefined is pretty unusual.

I agree, the implementation for stat.py and _stat.c is pretty unusual, probably for some unknown to us historical reason.  This is something to look into, but not as part of this issue.

I'll file a new issue for this.
